tensorflow/core/kernels/data/window_dataset_op.cc

```diff
@@ -103,8 +103,16 @@ class WindowDatasetOp : public UnaryDatasetOpKernel {
       if (n == kInfiniteCardinality || n == kUnknownCardinality) {
         return n;
       }
-      return n / window_shift_ +
-             (n % window_shift_ == 0 || drop_remainder_ ? 0 : 1);
+      int64 cardinality = 0;
+      if (drop_remainder_) {
+        while (1 + (n - 1) / window_stride_ >= window_size_) {
```

Comment:
Good catch. There is a closed-form (i.e. no iteration required) for this: `max(0, (n - ((window_size - 1) * window_stride + 1)) / window_shift + 1)`

Reply:
@jsimsa Yeah, this is more efficient than the iteration solution. There is a special case where `n = 7, window_size = 8, window_stride = 1, window_shift = 3`. Its cardinality should be `0` instead of `1`. Would the following form work?
```
int64 rest_elements = n - ((window_size_ - 1) * window_stride_ + 1);
cardinality = rest_elements < 0 ? 0 : rest_elements / window_shift_ + 1;
```

Reply:
The looks reasonable. I would add a comment there that explain the logic behind the formula:

Compute `rest_elements`, the number of elements after the last element of the initial window.
If it is negative, we know that the cardinality is 0. Otherwise, it will be the number of valid shifts over the `rest_elements`.

Reply:
The explanation is great! I add it to the code via https://github.com/tensorflow/tensorflow/pull/27934/commits/03616ab84bf9e4d0f54beff8a0b20b3ab86ee87f.